File: bindings/python/perf/multi/perf_multi_dealer_router_client.py
import sys
import time
from contextlib import ExitStack

import zlink

from perf_multi_common import (
    apply_multi_socket_options,
    benchmark_run_id,
    extract_metric_payload,
    is_active_message,
    latency_ns_from_message,
    new_payload,
    parse_client_args,
    perf_client_context,
    print_result_lines,
    recv_nonblocking,
    resolve_multi_connect_ready_timeout_ms,
    result_metrics,
    safe_poll,
    send_nonblocking,
    stamp_payload,
    wait_monitor_event,
)


def main(argv=None):
    args = parse_client_args(argv or sys.argv[1:], pattern="dealer_router")
    run_id = benchmark_run_id()
    payloads = [new_payload(args.msg_size) for _ in range(args.clients)]
    # C perf_multi_client_helpers.hpp run_echo_window_round_robin: the echo
    # client is purely deadline-driven (no wire stop token); awaiting_reply /
    # send_pending gate inflight==1 per socket.
    waiting_reply = [False] * args.clients
    send_pending = [True] * args.clients
    received = 0
    latencies = []
    seq = 0

    with perf_client_context() as ctx:
        sockets = [zlink.DealerSocket(ctx) for _ in range(args.clients)]
        try:
            with ExitStack() as stack:
                monitors = []
                for index, sock in enumerate(sockets):
                    sock.set_routing_id(f"CLIENT-{index}".encode("ascii"))
                    monitor = stack.enter_context(
                        sock.monitor_open(zlink.MonitorEventMask.CONNECTION_READY)
                    )
                    apply_multi_socket_options(sock)
                    sock.connect(args.endpoint)
                    monitors.append(monitor)
                for monitor in monitors:
                    wait_monitor_event(
                        monitor,
                        zlink.MonitorEventMask.CONNECTION_READY,
                        timeout_ms=resolve_multi_connect_ready_timeout_ms(),
                    )

                active_deadline = time.perf_counter() + args.duration
                recv_storage = [zlink.Received() for _ in sockets]
                with zlink.Poller() as poller:
                    for sock in sockets:
                        poller.add_socket(
                            sock,
                            zlink.PollEventFlag.POLLIN | zlink.PollEventFlag.POLLOUT,
                        )
                    # C run_echo_window_round_robin: loop until the steady
                    # deadline (no stop token). Each iteration sends pending
                    # sockets, then polls (-1, signal-driven) only the
                    # sockets with interest and dispatches only ready ones.
                    while time.perf_counter() < active_deadline:
                        for index, current_sock in enumerate(sockets):
                            if waiting_reply[index] or not send_pending[index]:
                                continue
                            seq += 1
                            payload = stamp_payload(
                                payloads[index],
                                phase=1,
                                run_id=run_id,
                                seq=seq,
                            )
                            if send_nonblocking(current_sock, payload):
                                waiting_reply[index] = True
                                send_pending[index] = False

                        # Build poll interest per C: POLLIN when awaiting a
                        # reply, else POLLOUT when a send is pending.
                        interest = [
                            i
                            for i in range(len(sockets))
                            if waiting_reply[i] or send_pending[i]
                        ]
                        if not interest:
                            for i in range(len(sockets)):
                                if not waiting_reply[i]:
                                    send_pending[i] = True
                            continue

                        # C uses perf_socket_poll(..., -1). The Poller has all
                        # sockets registered for POLLIN|POLLOUT; -1 wait is
                        # signal-driven and the result identifies ready
                        # sources for dispatch.
                        events = safe_poll(poller, -1)
                        if not events:
                            continue
                        for event in events:
                            current_sock = event.socket
                            try:
                                index = sockets.index(current_sock)
                            except ValueError:
                                continue
                            ev = int(event.events)
                            if (
                                ev & int(zlink.PollEventFlag.POLLOUT)
                                and not waiting_reply[index]
                                and send_pending[index]
                            ):
                                seq += 1
                                payload = stamp_payload(
                                    payloads[index],
                                    phase=1,
                                    run_id=run_id,
                                    seq=seq,
                                )
                                if send_nonblocking(current_sock, payload):
                                    waiting_reply[index] = True
                                    send_pending[index] = False
                            if not (ev & int(zlink.PollEventFlag.POLLIN)):
                                continue
                            while True:
                                msg = recv_nonblocking(
                                    current_sock,
                                    storage=recv_storage[index],
                                )
                                if msg is None:
                                    break
                                with msg:
                                    parts = msg.to_bytes_list()
                                waiting_reply[index] = False
                                data = extract_metric_payload(parts)
                                header = latency_ns_from_message(data) if data else None
                                # C: every matched header counts (++local_recv,
                                # → recv_count); latency only added when not
                                # clock-skewed (sent_ts_ns>0 && now>=sent_ts),
                                # halved for the round trip.
                                if data and is_active_message(
                                    data,
                                    expected_msg_size=args.msg_size,
                                    run_id=run_id,
                                ):
                                    received += 1
                                    if header is not None:
                                        latencies.append(header / 2.0)
                                if time.perf_counter() < active_deadline:
                                    send_pending[index] = True
                if received == 0:
                    raise RuntimeError(
                        "multi dealer-router benchmark did not receive any active reply"
                    )
                metrics = result_metrics(
                    count=received,
                    msg_size=args.msg_size,
                    elapsed_s=args.duration,
                    latencies_ns=latencies,
                    bandwidth_multiplier=2.0,
                )
                print_result_lines(
                    "MULTI_DEALER_ROUTER", args.transport, args.msg_size, metrics
                )
        finally:
            for sock in sockets:
                try:
                    sock.close()
                except Exception as exc:
                    print(f"[perf] close failed: {exc}", file=sys.stderr)


if __name__ == "__main__":
    main()

File: bindings/node/perf/multi/perf_multi_dealer_router_client.ts
// SPDX-License-Identifier: MPL-2.0

'use strict';

const readline = require('node:readline');
const zlink = require('../../dist/canonical');
const {
  createMetricCollector,
  createPayload,
  createRunId,
  decodeMetricHeader,
  currentEpochNs,
  sleepImmediate,
  summarizeMetrics,
  stampPayload
} = require('../common/perf_metrics');
const { parseMultiArgs } = require('./perf_multi_common');
const {
  POLLIN,
  POLLOUT,
  recvNoWait,
  trySocketSend,
  waitForConnectionReady
} = require('./perf_multi_runtime');

async function main() {
  const options = parseMultiArgs(process.argv.slice(2));
  const ctx = new zlink.Context();
  const dealers = [];
  const payloads = [];
  const waiting = [];
  const sendPending = [];
  const poller = new zlink.Poller();

  try {
    for (let i = 0; i < options.clients; i += 1) {
      const dealer = new zlink.DealerSocket(ctx);
      dealer.setRoutingId(zlink.RoutingId.fromBytes(Buffer.from(`CLIENT-${i}`, 'ascii')));
      dealers.push(dealer);
      payloads.push(createPayload(options.msgSize));
      waiting.push(false);
      sendPending.push(false);
    }
    for (let i = 0; i < dealers.length; i += 1) {
      await waitForConnectionReady(dealers[i], () => dealers[i].connect(options.endpoint));
      poller.addSocket(dealers[i], POLLIN | POLLOUT, i);
    }
    console.log(`CLIENT_READY,${options.msgSize}`);

    const rl = readline.createInterface({ input: process.stdin, crlfDelay: Infinity });
    for await (const line of rl) {
      if (line !== `START,${options.msgSize}`) {
        continue;
      }

      const runId = createRunId(1);
      const activeStartNs = currentEpochNs();
      const activeStopNs = activeStartNs + BigInt(Math.floor(options.duration * 1_000_000_000));
      const collector = createMetricCollector({
        runId,
        msgSize: options.msgSize,
        activeStartNs,
        activeStopNs,
        roundTrip: true
      });
      let seq = 1n;

      const drainReply = (index) => {
        let progressed = false;
        while (true) {
          const echoed = recvNoWait(dealers[index]);
          if (!echoed) {
            break;
          }
          waiting[index] = false;
          collector.record(
            decodeMetricHeader(echoed.parts[0].data()),
            currentEpochNs()
          );
          progressed = true;
        }
        return progressed;
      };

      while (currentEpochNs() < activeStopNs) {
        let progressed = false;
        for (let i = 0; i < dealers.length; i += 1) {
          if (waiting[i] || sendPending[i]) {
            continue;
          }
          stampPayload(payloads[i], { phase: 1, runId, msgSize: options.msgSize, seq });
          if (!trySocketSend(dealers[i], payloads[i])) {
            sendPending[i] = true;
            continue;
          }
          waiting[i] = true;
          seq += 1n;
          progressed = true;
        }
        for (let i = 0; i < dealers.length; i += 1) {
          progressed = drainReply(i) || progressed;
        }
        if (progressed) {
          continue;
        }

        const ready = poller.waitAll(poller.size, 25);
        if (ready.length === 0) {
          await sleepImmediate();
          continue;
        }
        for (const event of ready) {
          const index = event.userData;
          if (!Number.isInteger(index)) {
            continue;
          }
          if ((event.events & POLLOUT) !== 0) {
            sendPending[index] = false;
          }
          if ((event.events & POLLIN) !== 0) {
            drainReply(index);
          }
        }
      }

      const drainDeadline = Date.now() + 2000;
      while (waiting.some(Boolean) && Date.now() < drainDeadline) {
        let progressed = false;
        const ready = poller.waitAll(poller.size, 25);
        for (const event of ready) {
          const index = event.userData;
          if (!Number.isInteger(index)) {
            continue;
          }
          if ((event.events & POLLIN) !== 0) {
            progressed = drainReply(index) || progressed;
          }
          if ((event.events & POLLOUT) !== 0) {
            sendPending[index] = false;
          }
        }
        if (!progressed) {
          await sleepImmediate();
        }
      }

      const result = await collector.finish();
      for (const metricLine of summarizeMetrics(
        'MULTI_DEALER_ROUTER',
        options.transport,
        options.msgSize,
        result.latenciesNs,
        options.duration
      )) {
        console.log(metricLine);
      }
      break;
    }
  } finally {
    poller.close();
    for (const dealer of dealers) {
      dealer.close();
    }
    ctx.close();
  }
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});
